lib/hooks.js

```diff
@@ -192,67 +250,67 @@ function hookRunner (functions, runner, request, reply, cb) {
     }
   }
 
-  function handleResolve () {
-    next()
+  function handleResolve (newPayload) {
+    next(null, newPayload)
   }
 
   function handleReject (err) {
     if (!err) {
       err = new FST_ERR_SEND_UNDEFINED_ERR()
     }
 
-    cb(err, request, reply)
+    cb(err, request, reply, payload)
   }
 
   next()
 }
 
-function onSendHookRunner (functions, request, reply, payload, cb) {
+function preParsingHookRunner (functions, request, reply, cb) {
   let i = 0
 
-  function next (err, newPayload) {
-    if (err) {
-      cb(err, request, reply, payload)
+  function next (err, stream) {
+    if (reply.sent) {
       return
     }
 
-    if (newPayload !== undefined) {
-      payload = newPayload
+    if (typeof stream !== 'undefined') {
+      request[kRequestPayloadStream] = stream
     }
 
-    if (i === functions.length) {
-      cb(null, request, reply, payload)
+    if (err || i === functions.length) {
+      cb(err, request, reply)
       return
     }
 
     let result
     try {
-      result = functions[i++](request, reply, payload, next)
+      result = functions[i++](request, reply, request[kRequestPayloadStream], next)
     } catch (error) {
       next(error)
       return
     }
+
     if (result && typeof result.then === 'function') {
       result.then(handleResolve, handleReject)
     }
   }
 
-  function handleResolve (newPayload) {
-    next(null, newPayload)
+  function handleResolve (stream) {
+    next(null, stream)
   }
 
   function handleReject (err) {
     if (!err) {
       err = new FST_ERR_SEND_UNDEFINED_ERR()
     }
 
-    cb(err, request, reply, payload)
+    next(err)
```

Comment:
why does preParserHookRunner calls next and not cb when in handleReject. Other hookRunner call cb and not next, so it seems kind of fishy.

Reply:
Smells like a bug

Reply:
Seems like it. It also sounds tricky that this change does not break any tests. Can you add one for this case?

Reply:
@mcollina 

I checked it. As the next methods usually check first that err is an error or not, and then call cb, it means that next is calling itself one time more than  necessary.

So we could actually call directly cb instead of next,
See https://github.com/fastify/fastify/pull/4805/commits/10fdb5e2260dd9087fac28b75347b5f262ccbe5d